src/util/misc.rs

```diff
@@ -144,6 +145,31 @@ impl MessageSignature {
             compressed: self.compressed,
         })
     }
+
+    /// Verify that the signature signs the message and was signed by the given address.
+    ///
+    /// It returns [None] for addresses for which it is impossible to verify, like p2sh.
+    ///
+    /// To get the message hash from a message, use [signed_msg_hash].
+    pub fn verify_for_address<C: secp256k1::Verification>(
+        &self,
+        secp_ctx: &secp256k1::Secp256k1<C>,
+        address: &Address,
+        msg_hash: sha256d::Hash
+    ) -> Result<Option<bool>, secp256k1::Error> {
```

Comment:
Not sure if you thought about this but using a `bool` here could be suboptimal in some cases. Concretely, `bool` doesn't have a `#[must_use]` attribute, hence it is perfectly legal to write code that calls this function, handles the result but ignores the value inside `Ok`.

What about a method signature like:

```rust
    ) -> Result<(), MessageSignatureError> {
```

and add two variants to `MessageSignatureError`.

```rust
    /// The verification of this signature failed. The provided address DID NOT sign this message.
    Verification,
	/// The verification is impossible because the provided address is a script address.
    Impossible,
```
Doc-comments are subject to bikeshedding :)

Using a `Result` but not handling it will emit a warning at compile-time. Acting the Result will automatically fail the code if the verification fails - no hidden trap door :)

If you want to keep the `bool`, I would expect the function to be named `is_signed_by` or something similar that makes it obvious that this fn returns `bool`.

Reply:
Makes sense. I changed the signature to `is_signed_by_address` and the return value from `Result<Option<bool>,..>` to `Result<bool, ..>`. Which I think should make it safer to use.